dropshot_endpoint/src/error_store.rs

```diff
@@ -1,105 +1,164 @@
 // Copyright 2024 Oxide Computer Company
 
-//! Handle lists of errors in Dropshot.
+//! Handle lists of errors that occur while generating the proc macro.
+//!
+//! See the documentation of [`ErrorStore`] for more information.
 
 use std::cell::RefCell;
 
+/// Top-level struct that holds all errors encountered during the invocation of
+/// a proc macro.
+///
+/// ## Motivation
+///
+/// Dropshot's proc macros have several components that can all independently
+/// produce errors. We generally would like to make as much progress as possible
+/// within each of the components, because this tends to lead to better errors.
+///
+/// For example, the `endpoint` macro parses two components: the metadata, and
+/// the function signature.
+///
+/// * Failing to parse the metadata should not prevent us from parsing the
+///   function signature, and vice versa.
+/// * Within the signature, each type should be considered separately -- so an
+///   issue in an extractor type shouldn't stop us from parsing the return type.
+/// * But within a section, if there are errors, we should avoid producing the
+///   corresponding output -- because doing so can lead to confusing errors.
+///
+/// So the typical Rust pattern of returning on the first error isn't quite good
+/// enough for us, and what we need is:
+///
+/// * a hierarchical way to collect errors
```

Comment:
it's not clear to my why hierarchy is important or how it manifests to users... but this is great, and no need to block on my confusion!

Reply:
Basically the hierarchy is:

```
top level
|-> metadata
|-> params
      |-> request context
      |-> each extractor
      |-> return type
```

and errors in some node in the tree should not stop sibling nodes from being validated.